### PR TITLE
Identify tests over 0.2 flake factor and send a 'scarier' red message

### PR DESCRIPTION
**Pull request by Erin007**
[LP-1312](https://codedotorg.atlassian.net/browse/LP-1312), more follow up to [COE ](https://docs.google.com/document/d/1JhQ5TeNFt51qz_8fGgbW9GkksAPjeaPMEVbNn2kcfjw/edit?ouid=118428218617133301660&usp=docs_home&ths=true)

### CODE REVIEW

**Review by islemaster — APPROVED** [review]
:+1: Awesome!

File: bin/test_flakiness [hunk 1]
```diff
@@ -37,9 +40,15 @@ FileUtils.rm(TestFlakiness::CACHE_FILENAME) if options[:recalc]
 
 results = TestFlakiness.method(options[:method]).call
 results.to_a.sort_by(&:last).reverse[0..options[:limit]].map do |name, value|
-  result = [value, name]
```
Comment (uponthesun) [inline bin/test_flakiness]: Isn't this still needed below?

File: bin/test_flakiness [hunk 2]
```diff
@@ -37,9 +40,15 @@ FileUtils.rm(TestFlakiness::CACHE_FILENAME) if options[:recalc]
...
   if options[:reruns]
     result.insert(1, *TestFlakiness.recommend_reruns(TestFlakiness.test_flakiness[name]))
   end
-  puts result.join("\t")
+
+  if options[:flake_threshold]
```
Comment (uponthesun) [inline bin/test_flakiness]: Is `flake_threshold` different than `threshold`?

**Review by uponthesun — COMMENTED** [review]
Thanks for looking into this! Have you been able to test this yet / what does the output look like?

**Review by wjordan — COMMENTED** [review]
I like the general idea of extra warnings on flaky tests, here are some more suggestions/comments on the general approach:

- We already output a 'top 5' list of historically flakiest tests at the start of a test run, so I would expect another list of every test over a threshold to be somewhat redundant. Should we just make the existing top-5 list 'scarier' if any tests exceed this threshold?
- I wonder if this existing list gets ignored because it comes at the start of the test run rather than the end, so it ends up getting drowned out by the test results after it. Should we move this list to the end of the test run where it's closer to when the DotD will be considering their next followup actions?
- Since the `test_flakiness` script is only focused on historical test runs, it might be a little confusing to see a scary warning about flakiness in past runs when the test might actually pass its current run. It might be a bit better if we calculate test flakiness after the current run completes (if its API updates its history quickly enough), but might still be an issue even then (it might take several runs after a test is fixed before its calculated flakiness goes down below the threshold).

- As an alternative, we currently already log a test's flakiness on the line when it's auto-retried after failure. Could we change these auto-retry lines to be 'scarier' for historically-flaky tests, so that these failures are easier to spot and be warned about?
  - (related: it seems that flakiness reporting on these lines has been broken for a while with most showing up as `?`- I think I know the issue and will follow-up with a fix for this in a separate PR - #33787)

**Comment by Erin007** [comment]
> Thanks for looking into this! Have you been able to test this yet / what does the output look like?

How would I do that? Doesn't this have to be done in the context of a test run?

**Comment by uponthesun** [comment]
> How would I do that? Doesn't this have to be done in the context of a test run?

It being its own script in `bin` suggests to me that it should be runnable by itself, and the comment at the top says it uses data from Saucelabs, so it seems worth trying to just run it on the test machine. You can check out the file temporarily from this branch to test your changes also (I don't think there's anything on test which will automatically commit it?)

**Comment by Erin007** [comment]
I ran `bin/test_flakiness -x 0.2` on the test machine and it worked! 🎉

Screen Shot 2020-04-01 at 6 01 54 PM

**Comment by Erin007** [comment]
> * We already output a 'top 5' list of historically flakiest tests at the start of a test run, so I would expect another list of every test over a threshold to be somewhat redundant. Should we just make the existing top-5 list 'scarier' if any tests exceed this threshold?

I'm in agreement that having two lists is redundant.  Do we care about tests that _don't_ hit or exceed a 0.2 threshold (e.g. would it ever be important to know that a test had a flakiness of 1.5 versus 1.8)? If not, then maybe we just keep the new "scary" list that shows anything above the flake threshold and delete the one that shows top 5 flakiness regardless of whether they meet the threshold?

**Comment by Erin007** [comment]
> * I wonder if this existing list gets ignored because it comes at the start of the test run rather than the end, so it ends up getting drowned out by the test results after it. Should we move this list to the end of the test run where it's closer to when the DotD will be considering their next followup actions?
> * Since the `test_flakiness` script is only focused on historical test runs, it might be a little confusing to see a scary warning about flakiness in past runs when the test might actually pass its current run. It might be a bit better if we calculate test flakiness after the current run completes (if its API updates its history quickly enough), but might still be an issue even then (it might take several runs after a test is fixed before its calculated flakiness goes down below the threshold).

I moved the calculation and message to the end of the test run.  Now it's closer to the manual re-run warning, more likely to been seen by DotD considering follow-up and calculates test flakiness at the end of the run.  We'll still have the issue of the flakiness needing a few days to go down over time, but I think it might be ok for DotDs to be over-checking that a flaky test is being investigated by the cabal owners rather than never/seldom checking? Most folks have been including this kind of information in their end-of-shift emails and the past runs' flakiness output will all be in the history of the channel so a Dotd could check if a given test's flakiness was going up/down if they needed to.

**Pull request merged**